nemo/collections/common/prompts/formatter.py

```diff
@@ -0,0 +1,294 @@
+    # PromptFormatter.encode_dialog() ends with this role, it indicates a training example.
+    INFERENCE_ROLE = None
+
+    # Internal reserved field.
+    _REGISTERED_FORMATTERS = {}
+
+    def __init__(self, tokenizer: TokenizerSpec, defaults: list[dict] | None = None) -> None:
+        self.tokenizer = tokenizer
+        self._defaults = defaults
+        self._validate_defaults()
+
+    def __init_subclass__(cls, **kwargs) -> None:
+        ERR = "PromptFormatter subclass definition error:"
+        if cls.__name__ not in cls._REGISTERED_FORMATTERS:
+            for attr in ("REGISTER_NAME", "TEMPLATE", "INFERENCE_ROLE"):
+                assert (
+                    getattr(cls, attr, None) is not None
+                ), f"{ERR} PromptFormatter subclass {cls} did not define a class attribute {attr}"
+            cls._REGISTERED_FORMATTERS[cls.REGISTER_NAME] = cls
+        if "preamble" in cls.TEMPLATE:
+            assert (
+                len(cls.TEMPLATE["preamble"].get("slots", [])) == 0
+            ), f"{ERR} Slots are not allowed for preamble template, but we found: '{cls.TEMPLATE['preamble']}'"
+        for role in cls.get_roles():
+            template = cls.get_template(role)
+            for slot in cls.get_slots(role):
+                assert (
+                    _mangled(slot) in template
+                ), f"{ERR} Slot '{slot}' not found in template '{template}' for role '{role}'"
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def resolve(cls, name: str) -> Type["PromptFormatter"]:
+        if name not in cls._REGISTERED_FORMATTERS:
+            raise RuntimeError(
+                f"Unknown prompt formatter: '{name}' (known formats: {', '.join(cls._REGISTERED_FORMATTERS.keys())})"
+            )
+        return cls._REGISTERED_FORMATTERS[name]
+
+    @classmethod
+    @lru_cache(1)
+    def get_roles(cls) -> list[str]:
+        return list(cls.TEMPLATE.keys())
+
+    @classmethod
+    def get_slots(cls, role: str) -> dict[str, Modality]:
+        # returns a copy to avoid accidential mutation of a global object by the user
+        return cls.TEMPLATE[role].get("slots", {}).copy()
+
+    @classmethod
+    def get_template(cls, role: str) -> str:
+        return cls.TEMPLATE[role]["template"]
+
+    def get_default_dialog_slots(self) -> list[dict]:
+        """
+        Returns a list of dialog turns that can be used as a skeleton to fill with actual slot values.
+        If ``PromptFormatter`` was initialized with ``defaults`` argument, this method will return the
+        defaults. Otherwise, every slot is pre-filled with ``None``.
+        """
+        if self._defaults is not None:
+            return self._defaults.copy()
+        else:
+            return [
+                {"role": role, "slots": {slot: None for slot in self.get_slots(role)}}
+                for role in self.get_roles()
+                if role != self.INFERENCE_ROLE
```

Comment:
If I understand properly, this function provides an example of what the input dict should should like, which includes both the user's and the system's roles, right? If so, why do we exclude the INFERENCE_ROLE?

Reply:
it doesn't include the "assistant's response" (not to mix up with "system prompt"). it provides the user an easy way of creating a prompt for model inference. It doesn't make sense to have a default prompt for training as you need to fill the slots according to each training example